### ISSUE TITLE
关于SGL的初始化问题

### ISSUE THREAD

**Issue by 1197151063** [Coder-Yu/SELFRec#70]
俊良博士您好，我发现使用SELFRec运行SGL时无法复现到论文的结果，后来检查发现SGL的official implementation在初始化embedding时使用了标准差为0.01的正态分布，而SELFRec使用了Xavier均匀分布，经过测试后发现两种初始化方法对结果影响还挺大，想请教下这个初始化是trick吗？并且我发现LightGCN的Official implementation使用了标准差为0.1的正态分布作为初始化，其使用SGL的0.01分布时效果非常不理想，这种使用不同初始化的比较是否fair，感谢百忙中抽空回复！

**Comment by 1197151063** [comment]
不好意思，不清楚您已经博士毕业了！是我草率了w=w

**Comment by Coder-Yu** [comment]
我在做实验的时候所有方法统一使用的Xavier初始化， SGL也可以复现论文结果，你有选择合适的参数吗以及有设置足够大的迭代次数以使其收敛吗。一般来说，这几个方法使用Xavier初始化是可以达到好的效果的。使用不同的初始化确实算是一种trick，对比的时候应尽量使这些setting保持一致。

**Comment by 1197151063** [comment]
这是在Yelp2018上的部分training log，Recall@20比SGL的Official Implementation 低了 1%左右
我看Best Test Performance是在20个Epoch的时候取到，是因为我没多跑几个吗===
================================================================================
   SELFRec: A library for self-supervised recommendation.   
================================================================================
Graph-Based Baseline Models:
LightGCN   DirectAU   MF
----------------------------------------------------------------------------------------------------
Self-Supervised  Graph-Based Models:
SGL   SimGCL   SEPT   MHCN   BUIR   SelfCF   SSL4Rec   XSimGCL   NCL   MixGCF
================================================================================
Sequential Baseline Models:
SASRec
----------------------------------------------------------------------------------------------------
Self-Supervised Sequential Models:
CL4SRec   DuoRec   BERT4Rec
================================================================================
Please enter the model you want to run:SGL
Reading data and preprocessing...
  i = torch.LongTensor([coo.row, coo.col])
Model: SGL
Embedding Dimension: 64
Maximum Epoch: 20
Learning Rate: 0.001
Batch Size: 2048
Regularization Parameter: 0.0001
Specific parameters: n_layer:3  lambda:0.1  droprate:0.1  augtype:1  temp:0.2  
Training Set Size: (user number: 31668, item number 38048, interaction number: 1237259)
Test Set Size: (user number: 31668, item number 36073, interaction number: 324147)
================================================================================
Initializing and building model...
Training Model...
training: 1 batch 100 rec_loss: 0.6931027173995972 cl_loss 0.34818822145462036
training: 1 batch 200 rec_loss: 0.693098247051239 cl_loss 0.3484594523906708
training: 1 batch 300 rec_loss: 0.6930924654006958 cl_loss 0.34785929322242737
training: 1 batch 400 rec_loss: 0.693085789680481 cl_loss 0.34802210330963135
training: 1 batch 500 rec_loss: 0.6930767893791199 cl_loss 0.34773239493370056
training: 1 batch 600 rec_loss: 0.6930680274963379 cl_loss 0.34769997000694275
training: 2 batch 100 rec_loss: 0.6930544376373291 cl_loss 0.34873875975608826
training: 2 batch 200 rec_loss: 0.6930411458015442 cl_loss 0.34764230251312256
training: 2 batch 300 rec_loss: 0.6930269002914429 cl_loss 0.34815195202827454
training: 2 batch 400 rec_loss: 0.6930224895477295 cl_loss 0.34767213463783264
training: 2 batch 500 rec_loss: 0.6930112838745117 cl_loss 0.3482702672481537
training: 2 batch 600 rec_loss: 0.6929963827133179 cl_loss 0.3477550745010376
training: 3 batch 100 rec_loss: 0.692970871925354 cl_loss 0.34838375449180603
training: 3 batch 200 rec_loss: 0.6929556131362915 cl_loss 0.3479672372341156
training: 3 batch 300 rec_loss: 0.6929391026496887 cl_loss 0.347180038690567
training: 3 batch 400 rec_loss: 0.6929208040237427 cl_loss 0.3483850061893463
training: 3 batch 500 rec_loss: 0.692891538143158 cl_loss 0.3475845158100128
training: 3 batch 600 rec_loss: 0.6928857564926147 cl_loss 0.34716445207595825
training: 4 batch 100 rec_loss: 0.6928133368492126 cl_loss 0.3480513393878937
training: 4 batch 200 rec_loss: 0.6928001046180725 cl_loss 0.3482109010219574
training: 4 batch 300 rec_loss: 0.6927520036697388 cl_loss 0.3488308787345886
training: 4 batch 400 rec_loss: 0.6927405595779419 cl_loss 0.34788426756858826
training: 4 batch 500 rec_loss: 0.6926754713058472 cl_loss 0.3480498492717743
training: 4 batch 600 rec_loss: 0.6926414966583252 cl_loss 0.3479595482349396
training: 5 batch 100 rec_loss: 0.6925375461578369 cl_loss 0.34861528873443604
training: 5 batch 200 rec_loss: 0.6916712522506714 cl_loss 0.34787219762802124
training: 5 batch 300 rec_loss: 0.6916444301605225 cl_loss 0.34798067808151245
training: 5 batch 400 rec_loss: 0.6907984018325806 cl_loss 0.34851545095443726
training: 5 batch 500 rec_loss: 0.6918915510177612 cl_loss 0.3485218584537506
training: 5 batch 600 rec_loss: 0.6909828186035156 cl_loss 0.3475910425186157
training: 6 batch 100 rec_loss: 0.690631628036499 cl_loss 0.34905776381492615
training: 6 batch 200 rec_loss: 0.6908649206161499 cl_loss 0.34802934527397156
training: 6 batch 300 rec_loss: 0.6898835897445679 cl_loss 0.34787285327911377
training: 6 batch 400 rec_loss: 0.6886690258979797 cl_loss 0.3485729396343231
training: 6 batch 500 rec_loss: 0.6879217028617859 cl_loss 0.34868016839027405
training: 6 batch 600 rec_loss: 0.6872376799583435 cl_loss 0.3487940728664398
Evaluating the model...
Progress: [++++++++++++++++++++++++++++++++++++++++++++++++++]100%
------------------------------------------------------------------------------------------------------------------------

training: 20 batch 100 rec_loss: 0.06045796349644661 cl_loss 0.38324740529060364
training: 20 batch 200 rec_loss: 0.06038213148713112 cl_loss 0.3837977349758148
training: 20 batch 300 rec_loss: 0.05260005593299866 cl_loss 0.3833501636981964
training: 20 batch 400 rec_loss: 0.06306371092796326 cl_loss 0.3833737373352051
training: 20 batch 500 rec_loss: 0.05503180995583534 cl_loss 0.3825238049030304
training: 20 batch 600 rec_loss: 0.051803603768348694 cl_loss 0.38164806365966797
Evaluating the model...
Progress: [++++++++++++++++++++++++++++++++++++++++++++++++++]100%
------------------------------------------------------------------------------------------------------------------------
Real-Time Ranking Performance  (Top-20 Item Recommendation)
*Current Performance*
Epoch: 20, Hit Ratio:0.05868  |  Precision:0.03003  |  Recall:0.06689  |  NDCG:0.05497
*Best Performance* 
Epoch: 20, Hit Ratio:0.05868  |  Precision:0.03003  |  Recall:0.06689  |  NDCG:0.05497
------------------------------------------------------------------------------------------------------------------------
Testing...
Progress: [++++++++++++++++++++++++++++++++++++++++++++++++++]100%
Evaluating...
The result of SGL:
Top 20
Hit Ratio:0.05868
Precision:0.03003
Recall:0.06689
NDCG:0.05497

**Comment by 1197151063** [comment]
这个是Yelp2018上的部分Training log, Recall@20似乎要比SGL的Official Implementation 要低1%左右
我看早停在20个epoch时刚好test performance 最佳，似乎是因为我epoch没跑够===
代码里似乎Validation 和 Test是用的同一个dataset(或者说没有Validation)，想请教一下这个合不合理呢==

Please enter the model you want to run:SGL
Reading data and preprocessing...

  i = torch.LongTensor([coo.row, coo.col])
Model: SGL
Embedding Dimension: 64
Maximum Epoch: 20
Learning Rate: 0.001
Batch Size: 2048
Regularization Parameter: 0.0001
Specific parameters: n_layer:3  lambda:0.1  droprate:0.1  augtype:1  temp:0.2  
Training Set Size: (user number: 31668, item number 38048, interaction number: 1237259)
training: 1 batch 100 rec_loss: 0.6931027173995972 cl_loss 0.34818822145462036
training: 1 batch 200 rec_loss: 0.693098247051239 cl_loss 0.3484594523906708
training: 1 batch 300 rec_loss: 0.6930924654006958 cl_loss 0.34785929322242737
training: 1 batch 400 rec_loss: 0.693085789680481 cl_loss 0.34802210330963135
training: 1 batch 500 rec_loss: 0.6930767893791199 cl_loss 0.34773239493370056
training: 1 batch 600 rec_loss: 0.6930680274963379 cl_loss 0.34769997000694275
training: 2 batch 100 rec_loss: 0.6930544376373291 cl_loss 0.34873875975608826
training: 2 batch 200 rec_loss: 0.6930411458015442 cl_loss 0.34764230251312256
training: 2 batch 300 rec_loss: 0.6930269002914429 cl_loss 0.34815195202827454
training: 2 batch 400 rec_loss: 0.6930224895477295 cl_loss 0.34767213463783264
training: 2 batch 500 rec_loss: 0.6930112838745117 cl_loss 0.3482702672481537
training: 2 batch 600 rec_loss: 0.6929963827133179 cl_loss 0.3477550745010376
training: 3 batch 100 rec_loss: 0.692970871925354 cl_loss 0.34838375449180603
training: 3 batch 200 rec_loss: 0.6929556131362915 cl_loss 0.3479672372341156
training: 3 batch 300 rec_loss: 0.6929391026496887 cl_loss 0.347180038690567
training: 3 batch 400 rec_loss: 0.6929208040237427 cl_loss 0.3483850061893463
training: 3 batch 500 rec_loss: 0.692891538143158 cl_loss 0.3475845158100128
training: 3 batch 600 rec_loss: 0.6928857564926147 cl_loss 0.34716445207595825
training: 4 batch 100 rec_loss: 0.6928133368492126 cl_loss 0.3480513393878937
training: 4 batch 200 rec_loss: 0.6928001046180725 cl_loss 0.3482109010219574
training: 4 batch 300 rec_loss: 0.6927520036697388 cl_loss 0.3488308787345886
training: 4 batch 400 rec_loss: 0.6927405595779419 cl_loss 0.34788426756858826
training: 4 batch 500 rec_loss: 0.6926754713058472 cl_loss 0.3480498492717743
training: 4 batch 600 rec_loss: 0.6926414966583252 cl_loss 0.3479595482349396
training: 5 batch 100 rec_loss: 0.6925375461578369 cl_loss 0.34861528873443604
training: 5 batch 200 rec_loss: 0.6916712522506714 cl_loss 0.34787219762802124
training: 5 batch 300 rec_loss: 0.6916444301605225 cl_loss 0.34798067808151245
training: 5 batch 400 rec_loss: 0.6907984018325806 cl_loss 0.34851545095443726
training: 5 batch 500 rec_loss: 0.6918915510177612 cl_loss 0.3485218584537506
training: 5 batch 600 rec_loss: 0.6909828186035156 cl_loss 0.3475910425186157
training: 6 batch 100 rec_loss: 0.690631628036499 cl_loss 0.34905776381492615
training: 6 batch 200 rec_loss: 0.6908649206161499 cl_loss 0.34802934527397156
training: 6 batch 300 rec_loss: 0.6898835897445679 cl_loss 0.34787285327911377
training: 6 batch 400 rec_loss: 0.6886690258979797 cl_loss 0.3485729396343231
training: 6 batch 500 rec_loss: 0.6879217028617859 cl_loss 0.34868016839027405
training: 6 batch 600 rec_loss: 0.6872376799583435 cl_loss 0.3487940728664398
Evaluating the model...
training: 7 batch 100 rec_loss: 0.6842688322067261 cl_loss 0.3490992486476898
training: 7 batch 200 rec_loss: 0.6745395660400391 cl_loss 0.3508317768573761
training: 7 batch 300 rec_loss: 0.6744322180747986 cl_loss 0.35071632266044617
training: 7 batch 400 rec_loss: 0.6668941974639893 cl_loss 0.3524269163608551
training: 7 batch 500 rec_loss: 0.664889931678772 cl_loss 0.35234493017196655
training: 7 batch 600 rec_loss: 0.6564599871635437 cl_loss 0.3540952801704407
Evaluating the model...



Real-Time Ranking Performance  (Top-20 Item Recommendation)
Current Performance
Epoch: 7, Hit Ratio:0.01375  |  Precision:0.00704  |  Recall:0.0151  |  NDCG:0.01264
Best Performance 





Epoch: 20, Hit Ratio:0.05868  |  Precision:0.03003  |  Recall:0.06689  |  NDCG:0.05497
------------------------------------------------------------------------------------------------------------------------

Testing...
Progress: [++++++++++++++++++++++++++++++++++++++++++++++++++]100%
Evaluating...

The result of SGL:
Top 20
Hit Ratio:0.05868
Precision:0.03003
Recall:0.06689
NDCG:0.05497

**Comment by Coder-Yu** [comment]
SELFRec里面的这版本代码用的是batch内的user item作为对比学习的副样本，官方代码应该是用的所有user和item。你看下SGL论文的表5，跟你跑出来的结果实际上是吻合的。
validation set是需要的，我们是follow的SGL的做法把training set再采样一部分作为validation set。这个数据集的划分是直接用的他们的划分。validation set用过之后又merge到了training set里面。

**Comment by 1197151063** [comment]
谢谢解答！
